man/dmlog.1
.TH "DMLOG" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmlog \(en sends log to POSIX message queue
.SH SYNOPSIS
\f[B]dmlog\f[R] \-\-help
.PP
\f[B]dmlog\f[R] \-\-version
.PP
\f[B]dmlog\f[R] \-\-\f[B]node\f[R] \f[I]id\f[R] \-\-\f[B]message\f[R]
\f[I]string\f[R] [\-\-\f[B]logger\f[R] \f[I]name\f[R]]
[\-\-\f[B]level\f[R] \f[I]level\f[R]] [\-\-\f[B]error\f[R] \f[I]n\f[R]]
[\-\-\f[B]node\f[R] \f[I]id\f[R]] [\-\-\f[B]sensor\f[R] \f[I]id\f[R]]
[\-\-\f[B]target\f[R] \f[I]id\f[R]] [\-\-\f[B]observ\f[R] \f[I]id\f[R]]
[\-\-\f[B]source\f[R] \f[I]source\f[R]] [\-\-\f[B]debug\f[R]]
.SH DESCRIPTION
The \f[B]dmlog\f[R] utility forwards a log message to the POSIX message
queue of a \f[I]dmlogger(1)\f[R] instance.
Argument \f[CR]\-\-message\f[R] is mandatory.
Pass the name of the \f[I]dmlogger\f[R] or \f[I]dmrecv\f[R] instance to
send the log to through command\-line argument \f[CR]\-\-logger\f[R].
Alternatively, set environment variable \f[CR]DMLOGGER\f[R] to the name
of the logger.
The command\-line argument overwrites the environment variable.#
.PP
The default log level is \f[CR]info\f[R] (2).
Valid log levels are \f[CR]debug\f[R] (1), \f[CR]info\f[R] (2),
\f[CR]warning\f[R] (3), \f[CR]error\f[R] (4), \f[CR]critical\f[R] (5),
and \f[CR]user\f[R] (6).
The program terminates after log transmission.
.SH OPTIONS
.TP
\f[B]\-\-debug\f[R], \f[B]\-D\f[R]
Send log message of level \f[CR]debug\f[R] to logger.
.TP
\f[B]\-\-error\f[R], \f[B]\-e\f[R] \f[I]n\f[R]
DMPACK error code (>= 0).
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-level\f[R], \f[B]\-L\f[R] \f[I]level\f[R]
Log level, from \f[CR]debug\f[R] (1) to \f[CR]user\f[R] (6) The default
level is \f[CR]info\f[R] (2).
The argument may be an integer or log level name string.
.TP
\f[B]\-\-logger\f[R], \f[B]\-l\f[R] \f[I]name\f[R]
Name of logger instance and POSIX message queue.
If the command\-line argument is not passed and the environment variable
\f[CR]DMLOGGER\f[R] is not set, the log messages will not be transmitted
via POSIX message queue.
.TP
\f[B]\-\-message\f[R], \f[B]\-m\f[R] \f[I]string\f[R]
Log message (max.
512 characters).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Node id.
.TP
\f[B]\-\-observ\f[R], \f[B]\-O\f[R] \f[I]id\f[R]
Observation id (UUIDv4).
.TP
\f[B]\-\-sensor\f[R], \f[B]\-S\f[R] \f[I]id\f[R]
Sensor id.
.TP
\f[B]\-\-source\f[R], \f[B]\-Z\f[R] \f[I]source\f[R]
Name of log message source (sender).
.TP
\f[B]\-\-target\f[R], \f[B]\-T\f[R] \f[I]id\f[R]
Target id.
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print log to \f[I]stderr\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Log has been sent.
.TP
\f[B]1\f[R]
Failure.
Program execution failed.
.SH ENVIRONMENT
.TP
\f[B]DM_LOGGER\f[R]
Name of logger instance to send logs to.
.TP
\f[B]NO_COLOR\f[R]
Disable ANSI colour output.
.SH EXAMPLE
Send a log message to the POSIX message queue of logger
\f[CR]dmlogger\f[R]:
.IP
.EX
$ export DMLOGGER=dmlogger
$ dmlog \-N dummy\-node \-L warning \-m \(dqlow battery\(dq \-Z test \-V
2022\-12\-09T22:50:44.161000+01:00 [WARNING ] test \- low battery
.EE
.PP
The logger will receive the log message and insert it into the log
database (if the log level is greater or equal the minimum log level):
.IP
.EX
$ dmlogger \-N dummy\-node \-d log.db \-V
2022\-12\-09T22:50:44.161000+01:00 [WARNING] test \- low battery
.EE
.SH SEE ALSO
\f[I]dmlogger(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmlog.f90
! dmlog.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmlog
    !! Logging utility that sends log messages to a dmlogger instance if
    !! a valid logger is set through argument `--logger` or environment
    !! variable `DM_LOGGER`. The log message is printed to standard error if
    !! argument `--verbose` is passed. The default log level is `E_INFO`.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmlog'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    type :: app_type
        !! Command-line arguments.
        character(LOGGER_NAME_LEN) :: logger  = ' '     !! Name of logger instance (optional).
        logical                    :: debug   = .false. !! Forward debug messages via message queue (optional).
        logical                    :: verbose = .false. !! Print debug messages to stderr (optional).
    end type app_type

    class(logger_class), pointer :: logger

    integer        :: rc
    type(app_type) :: app
    type(log_type) :: log

    call dm_init()

    rc = read_args(app, log)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    logger => dm_logger_get_default()
    call logger%configure(name    = app%logger, & ! Name of logger process.
                          source  = APP_NAME,   & ! Default log source.
                          debug   = app%debug,  & ! Forward DEBUG messages via IPC.
                          ipc     = .true.,     & ! Enable IPC (if logger is set).
                          verbose = app%verbose)  ! Print logs to standard error.

    call dm_log_set(log, id=dm_uuid4(), timestamp=dm_time_now())
    call logger%log(log)
contains
    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS.
    ! **************************************************************************
    integer function read_args(app, log) result(rc)
        !! Reads command-line arguments.
        type(app_type), intent(out) :: app !! App type.
        type(log_type), intent(out) :: log !! Log type.

        type(arg_parser_class) :: parser

        ! Required and optional command-line arguments.
        call parser%add('logger',  short='l', type=ARG_TYPE_ID)      ! -l, --logger <id>
        call parser%add('level',   short='L', type=ARG_TYPE_LEVEL)   ! -L, --level <n>
        call parser%add('error',   short='e', type=ARG_TYPE_INTEGER) ! -e, --error <n>
        call parser%add('node',    short='N', type=ARG_TYPE_ID)      ! -N, --node <id>
        call parser%add('sensor',  short='S', type=ARG_TYPE_ID)      ! -S, --sensor <id>
        call parser%add('target',  short='T', type=ARG_TYPE_ID)      ! -T, --target <id>
        call parser%add('observ',  short='O', type=ARG_TYPE_UUID)    ! -O, --observ <uuid>
        call parser%add('source',  short='Z', type=ARG_TYPE_ID,     max_len=LOG_SOURCE_LEN)                   ! -Z, --source <id>
        call parser%add('message', short='m', type=ARG_TYPE_STRING, max_len=LOG_MESSAGE_LEN, required=.true.) ! -m, --message <string>
        call parser%add('debug',   short='D', type=ARG_TYPE_LOGICAL) ! -D, --debug
        call parser%add('verbose', short='V', type=ARG_TYPE_LOGICAL) ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('logger',  app%logger)
        call parser%get('level',   log%level, LL_INFO)
        call parser%get('error',   log%error)
        call parser%get('node',    log%node_id)
        call parser%get('sensor',  log%sensor_id)
        call parser%get('target',  log%target_id)
        call parser%get('observ',  log%observ_id)
        call parser%get('source',  log%source, APP_NAME)
        call parser%get('message', log%message)
        call parser%get('debug',   app%debug)
        call parser%get('verbose', app%verbose)

        rc = validate(log)
    end function read_args

    integer function validate(log) result(rc)
        !! Validates options and prints error messages.
        type(log_type), intent(inout) :: log !! Log type.

        rc = E_INVALID

        if (dm_string_has(app%logger) .and. .not. dm_id_is_valid(app%logger)) then
            call dm_error_out(rc, 'argument --logger is invalid')
            return
        end if

        if (.not. dm_error_is_valid(log%error)) then
            call dm_error_out(rc, 'argument --error is invalid')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
    end subroutine version_callback
end program dmlog
